File: frontend/src/api/savedSearches.ts
/**
 * Saved Searches & Pending Properties API
 */
import { apiClient } from './client';

// ── Saved Search Types ──────────────────────────────────────────────

export interface SavedSearch {
  id: string;
  user_id?: string;
  name: string;
  description?: string;
  portals: string[];
  property_type?: string;
  operation_type: string;
  city?: string;
  neighborhoods?: string[];
  province?: string;
  min_price?: number;
  max_price?: number;
  currency: string;
  min_area?: number;
  max_area?: number;
  min_bedrooms?: number;
  max_bedrooms?: number;
  min_bathrooms?: number;
  auto_scrape: boolean;
  is_active: boolean;
  last_executed_at?: string;
  total_executions: number;
  total_properties_found: number;
  pending_count: number;
  created_at: string;
  updated_at?: string;
}

export interface SavedSearchCreate {
  name: string;
  description?: string;
  portals: string[];
  property_type?: string;
  operation_type: string;
  city?: string;
  neighborhoods?: string[];
  province?: string;
  min_price?: number;
  max_price?: number;
  currency?: string;
  min_area?: number;
  max_area?: number;
  min_bedrooms?: number;
  max_bedrooms?: number;
  min_bathrooms?: number;
  auto_scrape?: boolean;
  is_active?: boolean;
}

export interface SavedSearchUpdate {
  name?: string;
  description?: string;
  portals?: string[];
  property_type?: string;
  operation_type?: string;
  city?: string;
  neighborhoods?: string[];
  province?: string;
  min_price?: number | null;
  max_price?: number | null;
  currency?: string;
  min_area?: number | null;
  max_area?: number | null;
  min_bedrooms?: number | null;
  max_bedrooms?: number | null;
  min_bathrooms?: number | null;
  auto_scrape?: boolean;
  is_active?: boolean;
}

export interface SavedSearchListResponse {
  total: number;
  skip: number;
  limit: number;
  items: SavedSearch[];
}

export interface SavedSearchExecuteResponse {
  success: boolean;
  search_id: string;
  search_name: string;
  total_found: number;
  new_properties: number;
  duplicates: number;
  scraped: number;
  pending: number;
  errors: Array<Record<string, unknown>>;
}

// ── Pending Property Types ──────────────────────────────────────────

export interface PendingProperty {
  id: string;
  saved_search_id: string;
  source_url: string;
  source: string;
  source_id?: string;
  title?: string;
  price?: number;
  currency?: string;
  thumbnail_url?: string;
  location_preview?: string;
  status: string;
  error_message?: string;
  property_id?: string;
  discovered_at: string;
  scraped_at?: string;
  updated_at?: string;
  saved_search_name?: string;
}

export interface PendingPropertyListResponse {
  total: number;
  skip: number;
  limit: number;
  items: PendingProperty[];
}

export interface PendingPropertyStats {
  total_pending: number;
  total_scraped: number;
  total_skipped: number;
  total_errors: number;
  by_search: Array<Record<string, unknown>>;
  by_portal: Array<Record<string, unknown>>;
}

export interface PendingPropertyScrapeResponse {
  success: boolean;
  scraped: number;
  errors: number;
  error_details: Array<Record<string, unknown>>;
}

export interface PendingPropertyActionResponse {
  success: boolean;
  message: string;
  pending_id: string;
  property_id?: string;
}

// ── API Client ──────────────────────────────────────────────────────

export const savedSearchesApi = {
  // Saved Searches
  list: async (skip: number = 0, limit: number = 50, active_only?: boolean): Promise<SavedSearchListResponse> => {
    const response = await apiClient.get<SavedSearchListResponse>('/saved-searches/', {
      params: { skip, limit, ...(active_only !== undefined && { active_only }) },
    });
    return response.data;
  },

  get: async (id: string): Promise<SavedSearch> => {
    const response = await apiClient.get<SavedSearch>(`/saved-searches/${id}`);
    return response.data;
  },

  create: async (data: SavedSearchCreate): Promise<SavedSearch> => {
    const response = await apiClient.post<SavedSearch>('/saved-searches/', data);
    return response.data;
  },

  update: async (id: string, data: SavedSearchUpdate): Promise<SavedSearch> => {
    const response = await apiClient.put<SavedSearch>(`/saved-searches/${id}`, data);
    return response.data;
  },

  delete: async (id: string): Promise<void> => {
    await apiClient.delete(`/saved-searches/${id}`);
  },

  execute: async (id: string, max_properties?: number): Promise<SavedSearchExecuteResponse> => {
    const response = await apiClient.post<SavedSearchExecuteResponse>(
      `/saved-searches/${id}/execute`,
      null,
      { params: max_properties !== undefined ? { max_properties } : {} }
    );
    return response.data;
  },

  toggle: async (id: string): Promise<SavedSearch> => {
    const response = await apiClient.post<SavedSearch>(`/saved-searches/${id}/toggle`);
    return response.data;
  },

  // Pending Properties
  listPending: async (
    skip: number = 0,
    limit: number = 50,
    search_id?: string,
    status_filter?: string,
    portal?: string
  ): Promise<PendingPropertyListResponse> => {
    const response = await apiClient.get<PendingPropertyListResponse>('/pending-properties/', {
      params: {
        skip,
        limit,
        ...(search_id && { search_id }),
        ...(status_filter && { status_filter }),
        ...(portal && { portal }),
      },
    });
    return response.data;
  },

  getPendingStats: async (): Promise<PendingPropertyStats> => {
    const response = await apiClient.get<PendingPropertyStats>('/pending-properties/stats');
    return response.data;
  },

  scrapeBatch: async (search_id?: string, limit: number = 50): Promise<PendingPropertyScrapeResponse> => {
    const response = await apiClient.post<PendingPropertyScrapeResponse>('/pending-properties/scrape', {
      ...(search_id && { search_id }),
      limit,
    });
    return response.data;
  },

  scrapeSingle: async (id: string): Promise<PendingPropertyActionResponse> => {
    const response = await apiClient.post<PendingPropertyActionResponse>(`/pending-properties/${id}/scrape`);
    return response.data;
  },

  skipPending: async (id: string): Promise<PendingPropertyActionResponse> => {
    const response = await apiClient.post<PendingPropertyActionResponse>(`/pending-properties/${id}/skip`);
    return response.data;
  },

  deletePending: async (id: string): Promise<void> => {
    await apiClient.delete(`/pending-properties/${id}`);
  },

  clearErrors: async (search_id?: string): Promise<{ success: boolean; cleared: number }> => {
    const response = await apiClient.post<{ success: boolean; cleared: number }>(
      '/pending-properties/clear-errors',
      null,
      { params: search_id ? { search_id } : {} }
    );
    return response.data;
  },
};

File: backend/app/api/v1/saved_searches.py
"""
Saved Searches API endpoints
"""
from typing import Optional
from uuid import UUID
from fastapi import APIRouter, Depends, HTTPException, status, Query
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy import select, func, delete

from app.database import get_db
from app.models.saved_search import SavedSearch
from app.models.pending_property import PendingProperty, PendingPropertyStatus
from app.schemas.saved_search import (
    SavedSearchCreate,
    SavedSearchUpdate,
    SavedSearchResponse,
    SavedSearchListResponse,
    SavedSearchExecuteResponse,
    ImportCardsRequest,
)
# from app.api.deps import get_current_user  # Temporarily disabled
# from app.models.user import User
from app.services.monitoring import MonitoringService


router = APIRouter()

# Temporary: no user filtering while auth is disabled
TEMP_USER_ID = None


@router.post("/", response_model=SavedSearchResponse, status_code=status.HTTP_201_CREATED)
async def create_saved_search(
    search_in: SavedSearchCreate,
    db: AsyncSession = Depends(get_db),
    # current_user: User = Depends(get_current_user),  # Temporarily disabled
):
    """
    Create a new saved search.

    Saved searches define the filters and portals to monitor for new properties.
    """
    # Create search
    search_data = search_in.model_dump()
    search_data['user_id'] = TEMP_USER_ID  # Temporarily hardcoded

    new_search = SavedSearch(**search_data)
    db.add(new_search)
    await db.commit()
    await db.refresh(new_search)

    # Add pending_count for response
    response = SavedSearchResponse.model_validate(new_search)
    response.pending_count = 0

    return response


@router.get("/", response_model=SavedSearchListResponse)
async def list_saved_searches(
    skip: int = Query(0, ge=0),
    limit: int = Query(50, ge=1, le=100),
    active_only: bool = Query(False, description="Only show active searches"),
    db: AsyncSession = Depends(get_db),
    # current_user: User = Depends(get_current_user),  # Temporarily disabled
):
    """
    List all saved searches for the current user.
    """
    # Build query (no user filter while auth is disabled)
    stmt = select(SavedSearch)

    if active_only:
        stmt = stmt.where(SavedSearch.is_active == True)

    # Get total count
    count_stmt = select(func.count()).select_from(
        stmt.subquery()
    )
    total_result = await db.execute(count_stmt)
    total = total_result.scalar() or 0

    # Get searches with pagination
    stmt = stmt.offset(skip).limit(limit).order_by(SavedSearch.created_at.desc())
    result = await db.execute(stmt)
    searches = result.scalars().all()

    # Get pending counts for each search
    items = []
    for search in searches:
        pending_count_stmt = select(func.count()).select_from(PendingProperty).where(
            PendingProperty.saved_search_id == search.id,
            PendingProperty.status == PendingPropertyStatus.PENDING
        )
        pending_result = await db.execute(pending_count_stmt)
        pending_count = pending_result.scalar() or 0

        response = SavedSearchResponse.model_validate(search)
        response.pending_count = pending_count
        items.append(response)

    return SavedSearchListResponse(
        total=total,
        skip=skip,
        limit=limit,
        items=items,
    )


@router.get("/{search_id}", response_model=SavedSearchResponse)
async def get_saved_search(
    search_id: UUID,
    db: AsyncSession = Depends(get_db),
    # current_user: User = Depends(get_current_user),  # Temporarily disabled
):
    """
    Get a saved search by ID.
    """
    stmt = select(SavedSearch).where(SavedSearch.id == search_id)
    result = await db.execute(stmt)
    search = result.scalar_one_or_none()

    if not search:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Búsqueda guardada no encontrada",
        )

    # Get pending count
    pending_count_stmt = select(func.count()).select_from(PendingProperty).where(
        PendingProperty.saved_search_id == search.id,
        PendingProperty.status == PendingPropertyStatus.PENDING
    )
    pending_result = await db.execute(pending_count_stmt)
    pending_count = pending_result.scalar() or 0

    response = SavedSearchResponse.model_validate(search)
    response.pending_count = pending_count

    return response


@router.put("/{search_id}", response_model=SavedSearchResponse)
async def update_saved_search(
    search_id: UUID,
    search_in: SavedSearchUpdate,
    db: AsyncSession = Depends(get_db),
    # current_user: User = Depends(get_current_user),  # Temporarily disabled
):
    """
    Update a saved search.
    """
    stmt = select(SavedSearch).where(SavedSearch.id == search_id)
    result = await db.execute(stmt)
    search = result.scalar_one_or_none()

    if not search:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Búsqueda guardada no encontrada",
        )

    # Update fields
    update_data = search_in.model_dump(exclude_unset=True)
    for field, value in update_data.items():
        setattr(search, field, value)

    await db.commit()
    await db.refresh(search)

    # Get pending count
    pending_count_stmt = select(func.count()).select_from(PendingProperty).where(
        PendingProperty.saved_search_id == search.id,
        PendingProperty.status == PendingPropertyStatus.PENDING
    )
    pending_result = await db.execute(pending_count_stmt)
    pending_count = pending_result.scalar() or 0

    response = SavedSearchResponse.model_validate(search)
    response.pending_count = pending_count

    return response


@router.delete("/{search_id}", status_code=status.HTTP_204_NO_CONTENT)
async def delete_saved_search(
    search_id: UUID,
    db: AsyncSession = Depends(get_db),
    # current_user: User = Depends(get_current_user),  # Temporarily disabled
):
    """
    Delete a saved search.

    This will also delete all associated pending properties (cascade).
    """
    stmt = select(SavedSearch).where(SavedSearch.id == search_id)
    result = await db.execute(stmt)
    search = result.scalar_one_or_none()

    if not search:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Búsqueda guardada no encontrada",
        )

    # Explicitly delete pending properties first to avoid async lazy-loading
    # issues with ORM cascade (MissingGreenlet in async context)
    await db.execute(
        delete(PendingProperty).where(PendingProperty.saved_search_id == search_id)
    )
    await db.delete(search)
    await db.commit()

    return None


@router.post("/{search_id}/execute", response_model=SavedSearchExecuteResponse)
async def execute_saved_search(
    search_id: UUID,
    max_properties: int = Query(100, ge=1, le=500, description="Max properties to discover per portal"),
    db: AsyncSession = Depends(get_db),
    # current_user: User = Depends(get_current_user),  # Temporarily disabled
):
    """
    Execute a saved search to discover new properties.

    This will:
    1. Scrape listing pages from configured portals
    2. Extract property URLs
    3. Check for duplicates
    4. Add new properties to pending queue (or auto-scrape if enabled)
    """
    stmt = select(SavedSearch).where(SavedSearch.id == search_id)
    result = await db.execute(stmt)
    search = result.scalar_one_or_none()

    if not search:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Búsqueda guardada no encontrada",
        )

    if not search.is_active:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="La búsqueda está desactivada",
        )

    # Execute search
    service = MonitoringService(db)
    results = await service.execute_search(search, max_properties=max_properties)

    return SavedSearchExecuteResponse(
        success=results['success'],
        search_id=search.id,
        search_name=search.name,
        total_found=results['total_found'],
        new_properties=results['new_properties'],
        duplicates=results['duplicates'],
        scraped=results['scraped'],
        pending=results['pending'],
        errors=results['errors'],
    )


@router.post("/{search_id}/import-cards", response_model=SavedSearchExecuteResponse)
async def import_cards(
    search_id: UUID,
    body: ImportCardsRequest,
    db: AsyncSession = Depends(get_db),
):
    """
    Import pre-scraped property cards into a saved search.

    Designed for the local scraper workflow: scraping happens on a local machine
    (residential IP, bypasses Cloudflare) and results are uploaded to the API.

    Deduplication by source_url, same as execute_search.
    """
    stmt = select(SavedSearch).where(SavedSearch.id == search_id)
    result = await db.execute(stmt)
    search = result.scalar_one_or_none()

    if not search:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Búsqueda guardada no encontrada",
        )

    service = MonitoringService(db)

    results = {
        'total_found': len(body.cards),
        'new_properties': 0,
        'duplicates': 0,
        'scraped': 0,
        'pending': 0,
        'errors': [],
    }

    for card_data in body.cards:
        card = card_data.model_dump()
        try:
            async with db.begin_nested():
                is_new, card_status = await service._process_discovered_property(
                    card=card,
                    search=search,
                    portal=_detect_portal(card_data.source_url, search.portals),
                )

            if is_new:
                results['new_properties'] += 1
                if card_status == 'scraped':
                    results['scraped'] += 1
                else:
                    results['pending'] += 1
            else:
                results['duplicates'] += 1

        except Exception as e:
            # Savepoint rollback is automatic — outer transaction stays valid
            results['errors'].append({
                'url': card_data.source_url,
                'error': str(e),
            })

    # Update search execution stats
    from datetime import datetime
    search.last_executed_at = datetime.utcnow()
    search.total_executions = (search.total_executions or 0) + 1
    search.total_properties_found = (search.total_properties_found or 0) + results['new_properties']

    await db.commit()

    return SavedSearchExecuteResponse(
        success=len(results['errors']) == 0,
        search_id=search.id,
        search_name=search.name,
        total_found=results['total_found'],
        new_properties=results['new_properties'],
        duplicates=results['duplicates'],
        scraped=results['scraped'],
        pending=results['pending'],
        errors=results['errors'],
    )


def _detect_portal(source_url: str, portals: list) -> str:
    """Detect portal from source URL, falling back to first portal in search."""
    url_lower = source_url.lower()
    if "zonaprop" in url_lower:
        return "zonaprop"
    if "argenprop" in url_lower:
        return "argenprop"
    if "remax" in url_lower:
        return "remax"
    if "mercadolibre" in url_lower:
        return "mercadolibre"
    # Fallback to first portal in the saved search
    if portals:
        return portals[0].lower() if hasattr(portals[0], 'lower') else str(portals[0]).lower()
    return "manual"


@router.post("/{search_id}/toggle", response_model=SavedSearchResponse)
async def toggle_saved_search(
    search_id: UUID,
    db: AsyncSession = Depends(get_db),
    # current_user: User = Depends(get_current_user),  # Temporarily disabled
):
    """
    Toggle the active state of a saved search.
    """
    stmt = select(SavedSearch).where(SavedSearch.id == search_id)
    result = await db.execute(stmt)
    search = result.scalar_one_or_none()

    if not search:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Búsqueda guardada no encontrada",
        )

    search.is_active = not search.is_active
    await db.commit()
    await db.refresh(search)

    # Get pending count
    pending_count_stmt = select(func.count()).select_from(PendingProperty).where(
        PendingProperty.saved_search_id == search.id,
        PendingProperty.status == PendingPropertyStatus.PENDING
    )
    pending_result = await db.execute(pending_count_stmt)
    pending_count = pending_result.scalar() or 0

    response = SavedSearchResponse.model_validate(search)
    response.pending_count = pending_count

    return response
